### PR TITLE
@W-10488954@: Implemented hardcoded rule skipping mechanism.

### PR DESCRIPTION
**Pull request by jfeingold35**
This PR does the following:
- Adds a mechanism for filtering specific rules from the results of a PMD execution.
- Hardcodes the specific rules in question to be "EagerlyLoadedSObjectDescribeResult".
- Adds `EVENTS` enum to use with emitting and receiving events.

### CODE REVIEW

File: src/lib/pmd/PmdEngine.ts [hunk 1]
```diff
@@ -304,6 +316,40 @@ abstract class BasePmdEngine extends AbstractRuleEngine {
...
+			for (const violation of result.violations) {
+				const ruleKey = violation.ruleName.toLowerCase();
+				// If the violation isn't for a skipped rule, add it to the filtered list.
+				if (!this.SKIPPED_RULES_TO_REASON_MAP.has(ruleKey)) {
+					filteredViolations.push(violation);
+				} else if (!alreadySkippedRules.has(ruleKey)) {
+					// If this is the first time we're skipping a violation for this particular rule, throw a verbose-only
+					// info-log about it.
+					uxEvents.emit(
+						EVENTS.INFO_VERBOSE,
```
Comment (jfeingold35) [inline src/lib/pmd/PmdEngine.ts:333]: The `EVENTS` enum was included so it could be used here.

File: src/lib/ScannerEvents.ts [hunk 2]
```diff
@@ -1,3 +1,12 @@
 import {EventEmitter} from 'events';
 
 export const uxEvents = new EventEmitter();
+
+export enum EVENTS {
```
Comment (jfeingold35) [inline src/lib/ScannerEvents.ts:5]: The `EVENTS` enum only contains the values that are currently relevant.

File: src/lib/pmd/PmdEngine.ts [hunk 3]
```diff
@@ -71,6 +71,13 @@ const HARDCODED_RULE_DETAILS: HardcodedRuleDetail[] = [
 
 
 abstract class BasePmdEngine extends AbstractRuleEngine {
+	/**
+	 * As per our internal policies, we want to avoid significantly changing output in minor releases, except for
+	 * high-severity security rules.
+	 * This Map is a temporary solution allowing us to silence new rules until they can be properly integrated.
+	 * @private
+	 */
+	private SKIPPED_RULES_TO_REASON_MAP: Map<string,string> = new Map([['eagerlyloadeddescribesobjectresult', 'Semantic version incompatible']]);
```
Comment (jfeingold35) [inline src/lib/pmd/PmdEngine.ts:80]: This is the new map of skipped rules.

File: src/lib/pmd/PmdEngine.ts [hunk 4]
```diff
@@ -172,15 +179,20 @@ abstract class BasePmdEngine extends AbstractRuleEngine {
...
-				violations = this.xmlToRuleResults(elements);
+				results = this.xmlToRuleResults(elements);
 			}
 		}
 
-		if (violations.length > 0) {
+		// For now, we're handling filtering by running all rules, and then removing results for the skipped rules post-facto.
+		// This is acceptable currently, since we're only skipping rules that are functional but incompatible with our
+		// versioning schema.
+		results = this.filterSkippedRulesFromResults(results);
```
Comment (jfeingold35) [inline src/lib/pmd/PmdEngine.ts:189]: This is where the filtering happens.

File: src/lib/pmd/PmdEngine.ts [hunk 5]
```diff
@@ -304,6 +316,40 @@ abstract class BasePmdEngine extends AbstractRuleEngine {
 		};
 	}
 
+	private filterSkippedRulesFromResults(unfilteredResults: RuleResult[]): RuleResult[] {
```
Comment (jfeingold35) [inline src/lib/pmd/PmdEngine.ts:319]: This is the method that does the filtering.

File: test/lib/pmd/PmdEngine.test.ts [hunk 6]
```diff
@@ -95,12 +96,101 @@ describe('Tests for BasePmdEngine and PmdEngine implementation', () => {
...
-			Sinon.assert.calledWith(uxSpy, 'warning-always', expectedConfigError);
-			Sinon.assert.calledWith(uxSpy, 'warning-always', expectedError);
-			Sinon.assert.calledWith(uxSpy, 'warning-always', expectedSuppressedViolation);
+			Sinon.assert.calledWith(uxSpy, EVENTS.WARNING_ALWAYS, expectedConfigError);
+			Sinon.assert.calledWith(uxSpy, EVENTS.WARNING_ALWAYS, expectedError);
+			Sinon.assert.calledWith(uxSpy, EVENTS.WARNING_ALWAYS, expectedSuppressedViolation);
 		});
 	});
 
+	describe('#filterSkippedRulesFromResults()', () => {
```
Comment (jfeingold35) [inline test/lib/pmd/PmdEngine.test.ts:105]: Covered cases:
- Removing one violation from a result
- Removing all violations for a result and subsequently dropping the result altogether
- Confirming that the event fires as expected.

**Review by rmohan20 — APPROVED** [review]

**Pull request merged**